Add VERSIONINFO resource

nkh0472 opened the pull request:
Display version info and file name under Windows.

Tested through latest Mingw-w64.

nkh0472 committed:
Add VERSIONINFO resource

https://docs.microsoft.com/en-us/windows/win32/menurc/versioninfo-resource

nkh0472 committed:
Correct FILETYPE

nkh0472 commented on fdkaac.rc:
```diff
@@ -1,3 +1,37 @@
+#define VER_VERSION 1,0,2,0
+//The version parameter is two 32-bit integers, defined by four 16-bit integers.
+#define VER_VERSION_STR "1.0.2"
+#define VER_LEGALCOPYRIGHT_STR "(c) nu774 2013-2021"
```
I'm not sure if you like this copyright notice, you can modify it or delete this line and line 26 if you don't like it.

nu774 commented:
Thanks for the patch.
I think it would be better if version number is automatically picked up from version.h.
Is it possible?

nkh0472 committed:
Move version defination to version.h

nkh0472 committed:
Moved version definition from fdkaac.rc

nkh0472 committed:
Merge pull request #1 from nkh0472/patch-2

Move version definition to version.h

nkh0472 commented on fdkaac.rc:
```diff
@@ -1,3 +1,33 @@
+	BLOCK "040904E4"
+	{
+		VALUE "FileDescription", "Command line frontend for libfdk-aac encoder"
+		VALUE "FileVersion", VER_VERSION_STR
```
I don't know why I can't use that variable `fdkaac_version`. 
```suggestion
		VALUE "FileVersion", fdkaac_version
```
That will cause 
```
$ windres fdkaac.rc -o fdkaac.o
E:\Etc\HUGE\msys64\mingw64\bin\windres.exe: fdkaac.rc:21: syntax error
```

nkh0472 commented:
> I think it would be better if version number is automatically picked up from version.h.

Now the definition of the version number has been moved to `version.h`.

nkh0472 committed:
Merge var VER_VERSION_STR and fdkaac_version

nkh0472 committed:
Merge var VER_VERSION_STR and fdkaac_version

nkh0472 committed:
Add \0 at the end of VERSIONINFO strings

When defining the version information string, all strings must be NULL-terminated strings, so add \0 at the end of the string.

nkh0472 committed:
Add \0 at the end of VERSIONINFO strings

nu774 commented:
Did you try re-running autoreconf ?
It now fails because configure.ac also picks up version information from version.h, and now it is different.
The first lines of configure.ac has to be edited to something like this.
```
m4_define([VERSION_STR],m4_esyscmd([grep VER_VERSION_STR version.h]))
changequote({{,}})dnl
m4_define({{XX_VERSION}},m4_bregexp(VERSION_STR,{{.*VER_VERSION_STR "\(.+\).0".*}},{{\1}}))
```
Also, I'm skeptical of the last two commits which add explicit NUL at the end of the string.
Are they *really* needed? Does that make difference on the generated file?

nkh0472 commented:
> Did you try re-running autoreconf ?
> It now fails because configure.ac also picks up version information from version.h, and now it is different.
> The first lines of configure.ac has to be edited to something like this.

For the last four commit, I only tested `make` and the output of the complied fdkaac.exe.

> Also, I'm skeptical of the last two commits which add explicit NUL at the end of the string.
> Are they _really_ needed? Does that make difference on the generated file?

I only saw such a sentence on [a Chinese blog page](http://www.cppblog.com/luqingfei/archive/2010/09/21/127239.html) today. Although the author did not give a reference, but in [Microsoft's official example](https://docs.microsoft.com/en-us/windows/win32/menurc/versioninfo-resource), all of the strings are written like this. So I think it's better to keep the writing consistent with the official documents.

Although there is no `\0` as the end in this commit before, and the file properties can be displayed normally in my Windows 10; But I'm worried about whether this will cause compatibility or other problems on lower version systems or another compiler.

nkh0472 committed:
Fix a problem caused by VERSION string change

```
$ autoreconf -i
/usr/bin/m4:configure.ac:1: Warning: excess arguments to builtin `m4_define' ignored
autom4te: error: /usr/bin/m4 failed with exit status: 1
aclocal-1.16: error: /usr/bin/autom4te failed with exit status: 1
autoreconf: error: aclocal failed with exit status: 1
```
https://github.com/nu774/fdkaac/pull/47#issuecomment-862961775

nu774 commented:
In C programming language, any string literal is terminated by a NUL, so you generally don't need them unless you need double NULs.
If you are unsure, you should at least check if output of resource compiler (windres) is actually different or not.
To begin with, don't push code if you are unsure. This is not your own repo.

nkh0472 commented:
> If you are unsure, you should at least check if output of resource compiler (windres) is actually different or not.

They are the same. But I still think it's better to follow the example of the official document.

> To begin with, don't push code if you are unsure. This is not your own repo.

I'm very sorry for the problem I caused. I will remember your advice.
